bionic/core/task_execution.py

```diff
@@ -16,46 +32,78 @@
+    if we have any dependent tasks that we plan to run in a separate process, we can
+    go ahead and start them; there may be more work to do on this task, but it will
+    have to be done in that separate process, because its results can't be serialized
+    and transmitted. (On the other hand, if we have a dependent task to run *in this
+    same process*, we'll want to bring this task to the "cached" state instead.) As
+    with (a), for a task with non-persistable output, this milestone is reached as
+    soon as we compute its provenance; for a task with persistable output, it's
+    reached only when the task is computed and its output is cached.
+
+    Thus, the operational definition of "primed" depends on whether the task's output is
+    persistable or not. If so, "primed" is equivalent to "cached"; if not, "primed"
+    represents a preliminary stage of work which happens to be sufficient for many
+    applications.
     """
 
-    def __init__(self, state, is_needed_in_memory):
+    def __init__(self, state):
         self.state = state
         self.stage = EntryStage.NEW
         self.future = None
         self.results_by_dnode = None
 
-        # This is initially set to None as we don't the entire dependency
-        # graph for entries at once. We create it only when it's computed.
-        self._dep_entries = None
+        # This is initially set to None to avoid eagerly recursing through the entire
+        # DAG. We set this once we start processing the entry.
+        self.dep_entries = None
 
-        # If an entry is needed in memory, it cannot be deferred. Only
-        # applicable to non-serializable entries, serializable entries
-        # are never deferred.
-        # NOTE An entry that is not needed in memory by one entry can
-        # be needed in memory by another entry. Depending on our
-        # execution order of such entities, this flag can change from
-        # `False` to `True` in between execution.
-        self.is_needed_in_memory = is_needed_in_memory
+        # These are initially set to False, but can be updated to True when we discover
+        # a requirement from another entry. However, they should never go from True to
+        # False!
+        self.is_requested_to_be_primed = False
+        self.is_requested_to_be_cached = False
 
     @property
     def is_cached(self):
         return self.results_by_dnode is not None or self.state.is_cached
 
-    def are_dep_entries_set(self):
-        return self._dep_entries is not None
-
-    def set_dep_entries(self, dep_entries):
-        assert not self.are_dep_entries_set()
-        self._dep_entries = dep_entries
+    @property
+    def is_primed(self):
+        if self.state.should_persist is None:
+            return False
+        if self.state.should_persist:
+            return self.state.is_cached
+        else:
+            return self.state.is_initialized
 
     def blocking_dep_entries(self):
```

Comment:
Thanks for adding the other TODO. As we discussed on the same call, can you also leave a TODO here to merge the logic for `blocking_dep_entries` and `all_dependencies_satisfy_requests`?